enhancements/ingress/haproxy-max-connections-tuning.md

```diff
@@ -0,0 +1,350 @@
+
+```sh
+$ oc patch ingresscontroller/default --type=merge --patch '{"spec":{"tuningOptions":{"maxConnections":-1}}}' -n openshift-ingress-operator
+```
+
+We can now extract the computed value from HAProxy's built-in stats
+socket:
+
+```sh
+$ oc rsh -n openshift-ingress <ROUTER-POD> bash -c 'echo "show info" | socat /var/lib/haproxy/run/haproxy.sock stdio' | grep Maxconn
+Maxconn: 524260
+```
+
+### Risks and Mitigations
+
+#### Increased Resource Usage
+
+HAProxy builds its data structures ahead of time. If you specify a
+large value for `spec.tuningOptions.maxConnections` then that memory
+is allocated up-front when the process starts. It is never released.
+It's possible that an administrator could set too high a value, given
+the node's configuration, causing other pods on the same node to
+become resource starved.
+
+Websocket connections or, more generally, long-lived connections can
+exacerbate memory usage for OpenShift router pods. As OpenShift router
+reloads, a new HAProxy process is created to run the new
+configuration. The current process will not terminate until the
+connections it is serving are all closed. In a busy cluster reloads
+could occur every 5 seconds and this has the potential to leave a
+long-tail of haproxy processes each consuming a significant amount of
+memory. This is particularly concerning when the auto-computed value
+(i.e., `-1`) is specified. In this scenario the additional memory used
+can be as much as 250MB per process.
+
+#### Unsupportable runtime limits
+
+An administrator can set a large value for
+`spec.tuningOptions.maxConnections` that cannot be satisfied because
+the computation taken by HAProxy doubles the asked-for-value to allow
+connections to queue up when the maximum has been reached. This is
+known internally to HAProxy as `maxsock`. Specifying
+`spec.tuniningOptions.maxConnections: 1048576` yields the following
+alert when the OpenShift router pod starts:
+
+```console
+$ oc logs -c router router-default-b696bd6cd-5qhqb
+  ...
+[ALERT] 095/124957 (30) : [/usr/sbin/haproxy.main()] FD limit (1048576) too low for maxconn=1048576/maxsock=2097208.
+Please raise 'ulimit-n' to 2097208 or more to avoid any trouble.
+This will fail in >= v2.3
+```
+
+Here we see HAProxy's computation as "maxsock = 2 * maxconn + 56". The
+additional 56 are based on internal HAProxy housekeeping requirements
+(e.g., stats port) and the set of configured frontend/backends in the
+`haproxy.config` file.
+
+There are two mitigation paths for this scenario:
+
+1. If you are setting extremely large values, always elect to use `-1`
+   (i.e., auto) and let HAProxy compute the value based on actual
+   ulimits within the running container, its own housekeeping
+   requirements, and the `frontend/backend/listen` entries specified
+   in `haproxy.config`.
+
+2. If you want an exact value then set a value that is improbably
+   large, heed the warning and the suggested fix, then configure a
```

Comment:
I'm a little confused by the wording here.  Is the "improbably large" value the one that the user actual wants to set?  Is heeding the warning and suggested fix the same step as configuring the tuned profile?